Python/optimizer_analysis.c

```diff
@@ -1,50 +1,1313 @@
+            if (new_inst.opcode == _LOAD_CONST_INLINE) {
+                Py_INCREF(val);
+            }
+            new_inst.operand = (uintptr_t)val;
+            break;
+        }
+        case STORE_FAST_MAYBE_NULL:
+        case STORE_FAST: {
+            _Py_UOpsSymbolicValue *value = PEEK(1);
+            _Py_UOpsSymbolicValue *new_local = sym_init_unknown(ctx);
+            if (new_local == NULL) {
+                goto error;
+            }
+            sym_copy_type_number(value, new_local);
+            GETLOCAL(oparg) = new_local;
+            STACK_SHRINK(1);
+            break;
+        }
+        case COPY: {
+            _Py_UOpsSymbolicValue *bottom = PEEK(1 + (oparg - 1));
+            STACK_GROW(1);
+            _Py_UOpsSymbolicValue *temp = sym_init_unknown(ctx);
+            if (temp == NULL) {
+                goto error;
+            }
+            PEEK(1) = temp;
+            sym_copy_type_number(bottom, temp);
+            break;
+        }
+
+        case POP_TOP: {
+            STACK_SHRINK(1);
+            break;
+        }
+
+        case PUSH_NULL: {
+            STACK_GROW(1);
+            _Py_UOpsSymbolicValue *null_sym =  sym_init_push_null(ctx);
+            if (null_sym == NULL) {
+                goto error;
+            }
+            PEEK(1) = null_sym;
+            break;
+        }
+
+        case _INIT_CALL_PY_EXACT_ARGS: {
+            _Py_UOpsSymbolicValue **__args_;
+            _Py_UOpsSymbolicValue *__self_or_null_;
+            _Py_UOpsSymbolicValue *__callable_;
+            _Py_UOpsSymbolicValue *__new_frame_;
+            __args_ = &stack_pointer[-oparg];
+            __self_or_null_ = stack_pointer[-1 - oparg];
+            __callable_ = stack_pointer[-2 - oparg];
+            // Store the frame symbolic to extract information later
+            assert(ctx->new_frame_sym.func == NULL);
+            ctx->new_frame_sym.func = __callable_;
+            ctx->new_frame_sym.self_or_null = __self_or_null_;
+            ctx->new_frame_sym.args = __args_;
+            __new_frame_ = _Py_UOpsSymbolicValue_New(ctx, NULL);
+            if (__new_frame_ == NULL) {
+                goto error;
+            }
+            stack_pointer[-2 - oparg] = (_Py_UOpsSymbolicValue *)__new_frame_;
+            stack_pointer += -1 - oparg;
+            break;
+        }
+
+        case _PUSH_FRAME: {
+            int argcount = oparg;
+            // TOS is the new frame.
+            STACK_SHRINK(1);
+            ctx->frame->stack_pointer = stack_pointer;
+            frame_info *frame_ir_entry = ir_frame_push_info(ctx, inst);
+            if (frame_ir_entry == NULL) {
+                goto error;
+            }
+
+            PyFunctionObject *func = extract_func_from_sym(&ctx->new_frame_sym);
+            if (func == NULL) {
+                goto error;
+            }
+            PyCodeObject *co = (PyCodeObject *)func->func_code;
```

Comment:
You can do this simpler -- a pointer to the code object is already stored in the uop as `operand`. (This is very recent, Mark added that when he wrote his peepholer.)

Reply:
Also, note that https://github.com/python/cpython/pull/114592 will change it to store the function, not the code object, as it needs access to the globals and builtins.

Reply:
Yeah. We need the changes in https://github.com/python/cpython/pull/114592 for function inlining too.

Reply:
This is a relic from the inlining - to inline we need the function and code object.

Reply:
This isn't safe for my purposes. By the time the optimizer gets hold of the code objects, the code objects may be freed, due to triggering GC or whatever by `Py_DECREF` or object allocations. So it may hold an invalid reference.

This only works for Mark's peepholer because the peepholer does not allocate objects or decref things.